File: Algorithms/graphs/DFS.py

def _in_bounds(grid, point):

    r, c = point
    return min(r, c) >= 0 and r < len(grid) and c < len(grid[0])

def get_neighbors(grid, point):

    r, c = point
    for rr, cc in [[r + 1, c], [r - 1, c], [r, c + 1], [r, c - 1]]:
        if _in_bounds(grid, (rr, cc)) and grid[rr][cc] == '.':
            yield rr, cc

def solve(grid, start, end):

    visited = []
    parents = {}

    stack = [start]

    while len(stack) > 0:

        cur = stack.pop()
        visited.append(cur)

        if cur == end:

            path = []
            node = cur

            while node != None:
                path.insert(0, node)
                node = parents.get(node)

            return path

        else:

            for n in get_neighbors(grid, cur):
                if n not in visited:
                    stack.append(n)
                    parents.update({n : cur})

    return None


def solve2(grid, cur, end, visited=None, path=None):

    if visited == None or path == None:
        return solve2(grid, cur, end, [], [])

    visited.append(cur)

    if cur == end:
        path.append(cur)
        return path
    else:
        for n in get_neighbors(grid, cur):
            if n not in visited:
                path.append(cur)
                new_path = solve2(grid, n, end, visited, path)
                if new_path:
                    return new_path
                else:
                    path.remove(cur)
    return None

grid = [
    ".####",
    ".#...",
    ".#.#.",
    "...#.",
    ".##.."
]

for p in solve(grid, (0,0), (4,4)):
    print(p)
print("------")
for p in solve2(grid, (0,0), (4,4)):
    print(p)

File: Algorithms/Graphs/DFS.js
"use strict";

Set.prototype.hasPoint = function(point) {

    for (let item of this.values()) {

        if (item.x == point.x && item.y == point.y) {

            return true;

        }

    }

    return false;

};

let inBounds = (grid, p) => Math.min(p.x, p.y) >= 0 && p.x < grid.length && p.y < grid[0].length;

function* getNeighbors(grid, point) {

    let {x, y} = point;

    for (let n of [{x: x + 1, y: y}, {x: x - 1, y: y}, {x: x, y: y + 1}, {x: x, y: y - 1}]) {

        if (inBounds(grid, n) && grid[n.x][n.y] == '.') {

            yield n;

        }

    }

}

let solve = (grid, start, end) => {

    let visited = new Set(),
        parents = new Map(),
        stack = [start];

    while (stack.length > 0) {

        let cur = stack.pop();
        visited.add(cur);

        if (cur.x == end.x && cur.y == end.y) {

            let path = [],
                node = cur;

            while (node != null) {

                path.unshift(node);
                node = parents.get(node);

            }

            return path;

        } else {

            for (let n of getNeighbors(grid, cur)) {

                if (!visited.hasPoint(n)) {

                    stack.push(n);
                    parents.set(n, cur);

                }

            }

        }

    }

    return null;

}

let solve2 = (grid, cur, end, visited = new Set(), path = []) => {

    visited.add(cur);

    if (cur.x == end.x && cur.y == end.y) {

        path.push(cur);

        return path;

    } else {

        for (let n of getNeighbors(grid, cur)) {

            if (!visited.hasPoint(n)) {

                path.push(cur);

                let newPath = solve2(grid, n, end, visited, path);

                if (newPath) {

                    return newPath;

                } else {

                  path.pop();

                }

            }

        }

    }

    return null;

}

let grid = [
    ".####",
    ".#...",
    ".#.#.",
    "...#.",
    ".##.."
];

for (let point of solve(grid, {x: 0, y: 0}, {x: 4, y: 4})) {

    console.log(point);

}

console.log("------");

for (let point of solve2(grid, {x: 0, y: 0}, {x: 4, y: 4})) {

    console.log(point);

}
